core/arch/riscv64/mangle.c

```diff
@@ -32,40 +32,170 @@
+            XINST_CREATE_load_int(dcontext, opnd_create_reg(DR_REG_A0), push_pc));
+    } else {
+        ASSERT(opnd_is_reg(push_pc));
+        reg_id_t push_pc_reg = opnd_get_reg(push_pc);
+        /* push_pc opnd is already pushed on the stack. */
+        PRE(ilist, instr,
+            INSTR_CREATE_ld(dcontext, opnd_create_reg(DR_REG_A0),
+                            OPND_CREATE_MEM64(DR_REG_SP, REG_OFFSET(push_pc_reg))));
+    }
+
+    PRE(ilist, instr,
+        INSTR_CREATE_sd(dcontext, OPND_CREATE_MEM64(DR_REG_SP, dstack_offs),
+                        opnd_create_reg(DR_REG_A0)));
+
+    dstack_offs += XSP_SZ;
+
+    /* Push FPRs. */
+    for (int i = 0; i < DR_NUM_FPR_REGS; i += 1) {
+        PRE(ilist, instr,
+            INSTR_CREATE_fsd(dcontext,
+                             opnd_add_flags(opnd_create_base_disp(
+                                                DR_REG_SP, DR_REG_NULL, 0,
+                                                dstack_offs + i * sizeof(reg_t), OPSZ_8),
+                                            DR_OPND_IMM_PRINT_DECIMAL),
+                             opnd_create_reg(DR_REG_F0 + i)));
+    }
+
+    dstack_offs += DR_NUM_FPR_REGS * XSP_SZ;
+
+    /* csrr a0, fcsr */
+    PRE(ilist, instr,
+        INSTR_CREATE_csrrs(dcontext, opnd_create_reg(DR_REG_A0),
+                           opnd_create_reg(DR_REG_X0),
+                           /* FIXME i#3544: Use register. */
+                           opnd_create_immed_int(0x003 /* fcsr */, OPSZ_12b)));
+
+    PRE(ilist, instr,
+        INSTR_CREATE_sd(dcontext, OPND_CREATE_MEM64(DR_REG_SP, dstack_offs),
+                        opnd_create_reg(DR_REG_A0)));
+
+    dstack_offs += XSP_SZ;
+
+    /* Keep mcontext shape. */
+    dstack_offs += (proc_num_simd_registers() * sizeof(dr_simd_t));
+
+    /* Restore the registers we used. */
+    PRE(ilist, instr,
+        INSTR_CREATE_ld(dcontext, opnd_create_reg(DR_REG_A0),
+                        OPND_CREATE_MEM64(DR_REG_SP, REG_OFFSET(DR_REG_A0))));
+
+    return dstack_offs;
 }
 
 void
 insert_pop_all_registers(dcontext_t *dcontext, clean_call_info_t *cci, instrlist_t *ilist,
-                         instr_t *instr, uint alignment _IF_AARCH64(bool out_of_line))
+                         instr_t *instr, uint alignment)
 {
-    /* FIXME i#3544: Not implemented */
-    ASSERT_NOT_IMPLEMENTED(false);
+    if (cci == NULL)
+        cci = &default_clean_call_info;
+    uint current_offs;
+    current_offs =
+        get_clean_call_switch_stack_size() - proc_num_simd_registers() * sizeof(reg_t);
+
+    PRE(ilist, instr,
+        INSTR_CREATE_ld(dcontext, opnd_create_reg(DR_REG_A0),
+                        OPND_CREATE_MEM64(DR_REG_SP, current_offs)));
+
+    /* csrw a0, fcsr */
+    PRE(ilist, instr,
+        INSTR_CREATE_csrrw(dcontext, opnd_create_reg(DR_REG_X0),
+                           opnd_create_reg(DR_REG_A0),
+                           opnd_create_immed_int(/*fcsr*/ 0x003, OPSZ_12b)));
```

Comment:
2nd time 0x003 is used: add a named constant for this file, even if temporary.  Put the same FIXME by its definition for longer-term solution.

Reply:
Done, also changed  `emit_utils.c` which uses `0x003` several times too. (I think a better way is to represent CSRs as registers than number literals, also added this to the comment).